### PR TITLE
Add tools/prepare_cache.py for offline GPT dataset cache preparation

### PR DESCRIPTION
**Pull request by asolergi-nv**
# What does this PR do ?
This PR adds a new offline cache-preparation entrypoint for GPTDataset-based training so dataset caches can be built ahead of time on a CPU-only node instead of forcing rank 0 to build them during training startup.

## What this introduces
  - Adds `tools/prepare_cache.py`
    - Reuses Megatron’s normal argument parsing/validation path so the tool can be launched with the same dataset-related training args
    - Computes train/valid/test sample targets via get_train_valid_test_num_samples() so cache keys match training behavior
    - Builds caches through the same GPTDataset + BlendedMegatronDatasetBuilder path used by training
    - Supports both:
        - --data-path + --split
        - per-split dataset definitions (--train-data-path, --valid-data-path, --test-data-path, --per-split-data-args-path)
    - Adds optional --prepare-cache-world-size to let a single-node prep run match the future training topology when sample counts depend on world size / DP size
  - Entry point parity changes
    - Updates pretrain_mamba.py so its GPTDatasetConfig construction matches GPT-relevant fields already present in pretrain_gpt.py, specifically:
        - multiple_validation_sets
        - full_validation
    - Includes a small formatting cleanup in pretrain_gpt.py

## Tests
Adds tests/unit_tests/data/test_prepare_cache.py covering:
  - cache preparation for blended datasets
  - cache preparation for per-split datasets
  - --prepare-cache-world-size normalization
  - unsupported-mode rejection
  - cache-hit rebuilds using:
      - dataloader_fast_cache_load=True
      - dataloader_defer_npy_index_mmap=True
  - verification that deferred/lazy-loaded cached datasets still return the same samples as the normal cache-hit path

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by asolergi-nv** [comment]
/ok to test [d72afb6](https://github.com/NVIDIA/Megatron-LM/pull/4080/commits/d72afb6d230e6aae26472be8f03f49330adf7e9c)

**Comment by asolergi-nv** [comment]
/ok to test [c87feda](https://github.com/NVIDIA/Megatron-LM/pull/4080/commits/c87feda19a58aff53f0986e85d3f2cacc9205bee)

**Comment by asolergi-nv** [comment]
/ok to test [2cd79f9](https://github.com/NVIDIA/Megatron-LM/pull/4080/commits/2cd79f9947c2c97dab7881a55446214d909641bb)

File: megatron/core/datasets/blended_dataset.py [hunk 1]
```diff
@@ -150,7 +150,10 @@ def _build_indices(self) -> Tuple[numpy.ndarray, numpy.ndarray]:
         else:
             cache_hit = False
 
-        if not path_to_cache or (not cache_hit and torch.distributed.get_rank() == 0):
+        if not path_to_cache or (
```
Comment (deepakn94) [inline megatron/core/datasets/blended_dataset.py:153]: Why do we need `and torch.distributed.get_rank() == 0` when we're already using `log_single_rank`?

**Review by deepakn94 — APPROVED** [review]

File: tools/prepare_cache.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    """Build the dataset caches for the plain GPTDataset path."""
+
+    _validate_prepare_cache_args(args)
+    ignored_flags = _disable_cache_load_only_flags(args)
+
+    unset_global_variables()
+    set_args(args)
+
+    try:
+        train_valid_test_num_samples = get_train_valid_test_num_samples()
```
Comment (maanug-nv) [inline tools/prepare_cache.py:175]: should call `update_train_iters()` here, to handle if only '--train-samples' is provided instead of'--train-iters'.

**Review by dimapihtar — APPROVED** [review]
LGTM. Thank you!

File: megatron/core/datasets/blended_dataset.py (hunk 1 above)
Comment (asolergi-nv) [reply megatron/core/datasets/blended_dataset.py:153]: I think you're overlooking the following lines of code. Could you expand them and take a look? I'm just adding the `torch.distributed.is_initialized()` check for protection.

File: tools/prepare_cache.py (hunk 2 above)
Comment (asolergi-nv) [reply tools/prepare_cache.py:175]: Done! Thanks!

**Comment by asolergi-nv** [comment]
/ok to test [c1641a7](https://github.com/NVIDIA/Megatron-LM/pull/4080/commits/c1641a774122857175b3b9fd329e1dc717f36dc6)

**Comment by asolergi-nv** [comment]
/ok to test [ec3cb71](https://github.com/NVIDIA/Megatron-LM/pull/4080/commits/ec3cb71b81e96bdd307f478a0d0913ced8e67123)

**Review by ericharper — APPROVED** [review]
